### PR TITLE
Algorithm: Calculating Product Sum from a Special Array with Nested Structures

### PR DESCRIPTION
**Pull request by Himanshutomar31**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Review by CaedenPH — APPROVED** [review]
The failing tests are not because of you

**Comment by Himanshutomar31** [comment]
> The failing tests are not because of you

Thanks for the review

**Comment by Himanshutomar31** [comment]
Hello @cclauss , would you kindly review my pull request and merge it? Thank you!

File: data_structures/arrays/product_sum.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+the product sum is x + 2 * (y + 3z).
+
+Example Input:
+[5, 2, [-7, 1], 3, [6, [-13, 8], 4]]
+Output: 12
+
+"""
+
+
+def product_sum(arr: list[int | list], depth: int) -> int:
```
[the remark this answers is not in the document]
Comment (Himanshutomar31) [reply data_structures/arrays/product_sum.py:23]: This algo has doc tests
`  Examples:
        >>> product_sum_array([1, 2, 3])
        6
        >>> product_sum_array([1, [2, 3]])
        11
        >>> product_sum_array([1, [2, [3, 4]]])
        47`

**Comment by Himanshutomar31** [comment]
Hi @cclauss , can you please review this PR?

File: data_structures/arrays/product_sum.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (cclauss) [reply data_structures/arrays/product_sum.py:23]: This function has no doctests.  If we copy just this function and put it into another project, it will have zero test coverage.  That is not acceptable.

[the remark this answers is not in the document]
Comment (Himanshutomar31) [reply data_structures/arrays/product_sum.py:23]: Hi, I have added the doctests for product_sum function, can you please review

File: data_structures/arrays/product_sum.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    Args:
+        arr: The array of integers and nested lists.
+        depth: The current depth level.
+
+    Returns:
+        int: The product sum of the array.
+
+    Examples:
+        >>> product_sum([1, 2, 3], 1)
```
Comment (cclauss) [inline data_structures/arrays/product_sum.py:39]: Please add tests for:
```
>>> product_sum([1, 2, 3], -1)
>>> product_sum([1, 2, 3], 0)
>>> product_sum([1, 2, 3], 7)
```
Can you please provide one or more URLs to discussions of this algorithm so that visitors can read up on where and when it is useful?  This should also give us access to correct answers for our tests.

File: data_structures/arrays/product_sum.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        6
+        >>> product_sum([-1, 2, [-3, 4]], 2)
+        8
+    """
+    total_sum = 0
+    for ele in arr:
+        if isinstance(ele, list):
+            total_sum += product_sum(ele, depth + 1)
+        else:
+            total_sum += ele
```
Comment (cclauss) [inline data_structures/arrays/product_sum.py:49]:
```suggestion
        total_sum += product_sum(ele, depth + 1) if isinstance(ele, list) else ele
```

File: data_structures/arrays/product_sum.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    Args:
+        array (List[Union[int, List]]): The array of integers and nested lists.
+
+    Returns:
+        int: The product sum of the array.
+
+    Examples:
+        >>> product_sum_array([1, 2, 3])
+        6
+        >>> product_sum_array([1, [2, 3]])
```
Comment (cclauss) [inline data_structures/arrays/product_sum.py:66]: Please add some tests with zero and negative integer and 1.5.

Comment (Himanshutomar31) [reply data_structures/arrays/product_sum.py:80]: Hi, I have updated the code as per the suggestions, can you please review

**Comment by Himanshutomar31** [comment]
Hi @cclauss  , can you please review this PR?

File: data_structures/arrays/product_sum.py (hunk 4 above)
Comment (cclauss) [reply data_structures/arrays/product_sum.py:80]: Still need `product_sum_array()` tests with zero and negative integer and 1.5.

Comment (Himanshutomar31) [reply data_structures/arrays/product_sum.py:80]: Hi, I have added the doctests for the product_sum_array() function, can you please review

**Comment by Himanshutomar31** [comment]
Hi @cclauss , can you please review this PR?

**Comment by Himanshutomar31** [comment]
Hi @cclauss  , can you please review this PR?

**Review by cclauss — APPROVED** [review]
